### train ANN Model



In [29]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten, Dense, Dropout

from tensorflow.keras import layers

In [4]:
from sklearn.model_selection import train_test_split

In [8]:
import pandas as pd

data = pd.read_csv("data_dropped.csv")

In [13]:
data.columns

Index(['Category', 'url_len', '@', '?', '-', '=', '.', '#', '%', '+', '$', '!',
       '*', ',', '//', 'abnormal_url', 'https', 'digits', 'letters',
       'Shortining_Service', 'having_ip_address'],
      dtype='object')

In [14]:
data

Category  url_len  @  ?  -  =  .  #  %  +  ...  !  *  ,  //  \
0              1       16  0  0  1  0  2  0  0  0  ...  0  0  0   0   
1              0       35  0  0  0  0  2  0  0  0  ...  0  0  0   0   
2              0       31  0  0  0  0  2  0  0  0  ...  0  0  0   0   
3              1       84  0  1  1  4  2  0  0  0  ...  0  0  0   1   
4              1      235  0  1  1  3  2  0  0  0  ...  0  0  0   1   
...          ...      ... .. .. .. .. .. .. .. ..  ... .. .. ..  ..   
651186         1       39  0  0  0  0  3  0  0  0  ...  0  0  0   0   
651187         1       44  0  0  2  0  2  0  0  0  ...  0  0  0   0   
651188         1       38  0  0  0  0  1  0  0  0  ...  0  0  0   0   
651189         1       45  0  0  0  0  2  0  0  0  ...  0  0  0   0   
651190         1       37  0  0  0  0  1  0  0  0  ...  0  0  0   0   

        abnormal_url  https  digits  letters  Shortining_Service  \
0                  0      0       0       13                   0   
1                  0      0       1       29                   0   
2                  0      0       1       25                   0   
3                  1      0       7       60                   0   
4                  1      0      22      199                   0   
...              ...    ...     ...      ...                 ...   
651186             0      0      12       21                   0   
651187             0      0       7       29                   1   
651188             0      0       3       30                   1   
651189             0      0       0       36                   0   
651190             0      0       0       33                   0   

        having_ip_address  
0                       0  
1                       0  
2                       0  
3                       0  
4                       0  
...                   ...  
651186                  0  
651187                  0  
651188                  0  
651189                  0  
651190                  0  

[651191 rows x 21 columns]

In [20]:
x_train, x_test, y_train, y_test = train_test_split(data.iloc[:,1:], data.iloc[:,0])

In [53]:
mlp = Sequential([
    Dense(128, activation = "relu"),
    Dense(64, activation="relu"),
    Dense(2, activation="softmax")
])

In [62]:
mlp = Sequential([
    Dense(128, activation = "sigmoid"),
    Dense(64, activation="sigmoid"),
    Dense(2, activation="softmax")
])

In [63]:
mlp.compile(optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"])

In [64]:
mlp.fit(x_train,y_train,epochs=10)

Epoch 1/10
15263/15263 [==============================] - 18s 1ms/step - loss: 0.3228 - accuracy: 0.8620
Epoch 2/10
15263/15263 [==============================] - 17s 1ms/step - loss: 0.2962 - accuracy: 0.8739
Epoch 3/10
15263/15263 [==============================] - 17s 1ms/step - loss: 0.2799 - accuracy: 0.8870
Epoch 4/10
15263/15263 [==============================] - 17s 1ms/step - loss: 0.2515 - accuracy: 0.9029
Epoch 5/10
15263/15263 [==============================] - 17s 1ms/step - loss: 0.2320 - accuracy: 0.9115
Epoch 6/10
15263/15263 [==============================] - 17s 1ms/step - loss: 0.2241 - accuracy: 0.9145
Epoch 7/10
15263/15263 [==============================] - 17s 1ms/step - loss: 0.2197 - accuracy: 0.9163
Epoch 8/10
15263/15263 [==============================] - 17s 1ms/step - loss: 0.2166 - accuracy: 0.9174
Epoch 9/10
15263/15263 [==============================] - 17s 1ms/step - loss: 0.2144 - accuracy: 0.9181
Epoch 10/10
15263/15263 [==============================

In [65]:
mlp.evaluate(x_test, y_test)

5088/5088 [==============================] - 5s 921us/step - loss: 0.2037 - accuracy: 0.9214


[0.20365652441978455, 0.9213749766349792]

### Test Result
With 2 layers of perceptron (128, 64), 10 epochs, both with 
- Rectified Linear Unit activation function, we got 92% accuracy on Train dataset and 92.4% accuracy on Test dataset. 
- Sigmoid function, we got 92.28% accuracy on Train dataset and 92.43% accuracy on Test Dataset.